### PR TITLE
Rebuild to provide mpi_f08 on all platforms

### PR DESCRIPTION
**Pull request by aradi**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [X] ~Reset the build number to `0` (if the version changed)~
* [X] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [X] Ensured the license file is being packaged.

After a longer odyssey, this PR is basically rebuilding MPICH using the latest (fixed versions) of the gfortran compiler. MPICH should now detect "F2008 support" (which is actually F2018 support) on all platforms, and consequently provide the `mpi_f08` interface everywhere.

The build script uses now `cp` instead of `ln` to put the `ISO_Fortran_binding.h` into the right search folder on OSX, as `cp` (in contrast to `ln`) fails reliably, if the file is not present.

### CODE REVIEW

**Comment by aradi** [comment]
@conda-forge-admin, please rerender

File: recipe/build-mpi.sh [hunk 1]
```diff
@@ -34,6 +34,11 @@ if [[ "$target_platform" == osx-* ]]; then
   fcdir=$($FC -print-search-dirs | awk '/install: /{print $2}')
   ccdir=$($CC -print-search-dirs | awk '/libraries: =/{print substr($2,2)}')
   ln -s $fcdir/include/ISO_Fortran_binding.h $ccdir/include
+  echo "=== DEBUG ==="
+  echo "CCDIR: ${ccdir}"
+  echo "Content of ${ccdir}"
+  echo "$(ls ${ccdir})"
```
Comment (dalcinl) [inline recipe/build-mpi.sh:40]: Maybe you meant the following?

```suggestion
  echo "Content of ${ccdir}/include"
  echo "$(ls ${ccdir}/include)"
```

Comment (aradi) [reply recipe/build-mpi.sh:40]: Hehe, yes, I just looked at results now, and yes, I meant exactly that. :smile:

**Comment by dalcinl** [comment]
The header file is there, indeed. What did you see from `cat config.log` after `./configure ...`? Did you look into it?

I'll try things out on baremetal to try to figure out what may be going on, although I'm not sure I'll be able to do so today.

**Comment by aradi** [comment]
@dalcinl I am debugging it step by step in the short breaks I have during my work today. If I find anything conclusive, I'll ping you.

File: recipe/build-mpi.sh [hunk 2]
```diff
@@ -33,7 +33,29 @@ if [[ "$target_platform" == osx-* ]]; then
   # Add gfortran internal header to clang include dir
   fcdir=$($FC -print-search-dirs | awk '/install: /{print $2}')
   ccdir=$($CC -print-search-dirs | awk '/libraries: =/{print substr($2,2)}')
-  ln -s $fcdir/include/ISO_Fortran_binding.h $ccdir/include
+  echo "=== DEBUG ==="
+  echo "cp ${fcdir}include/ISO_Fortran_binding.h ${ccdir}include/ISO_Fortran_binding.h"
+  cp ${fcdir}include/ISO_Fortran_binding.h ${ccdir}include/ISO_Fortran_binding.h
```
Comment (dalcinl) [inline recipe/build-mpi.sh:38]: You are missing a slash here. To be sure of getting it right just do

```suggestion
  echo "cp ${fcdir}/include/ISO_Fortran_binding.h ${ccdir}/include"
  cp ${fcdir}/include/ISO_Fortran_binding.h ${ccdir}/include
```

**Comment by dalcinl** [comment]
> OK, the conclusion is, that the `ISO_Fortran_binding.h` is simply missing on osx_arm64

That's not what I see when I install the packages manually in a conda environment. I believe your conclusion comes from the bug I reported above. Apply the fix and try again.

File: recipe/build-mpi.sh (hunk 2 above)
Comment (aradi) [reply recipe/build-mpi.sh:38]: Unfortunately, that won't help. (It was that way at the beginning, I've just removed it, to make sure, that it is not the double slash before include, which causes problems.)

If you look at the CI with the current build-mpi.sh, you see
```
2024-01-31T16:39:08.2427120Z cp /Users/runner/miniforge3/conda-bld/mpich-mpi_1706718393956/_build_env/lib/gcc/arm64-apple-darwin20.0.0/12.3.0/include/ISO_Fortran_binding.h /Users/runner/miniforge3/conda-bld/mpich-mpi_1706718393956/_build_env/lib/clang/16include/ISO_Fortran_binding.h
2024-01-31T16:39:08.2471840Z cp: /Users/runner/miniforge3/conda-bld/mpich-mpi_1706718393956/_build_env/lib/gcc/arm64-apple-darwin20.0.0/12.3.0/include/ISO_Fortran_binding.h: No such file or directory
```

The file is simply missing.

When you install the package manually, do you get GFortran 12 or 13 on your osx_arm64 platform. According to the CI output, GFortran 12.3 is used!

**Comment by dalcinl** [comment]
> (On all the other platforms, where it works, the compiler version is 13.2....)

Oh, that's a good point. I thought it was 12.3 everywhere, my bad. Now I see why I got such confusion. The recipe rerendering seems to have removed files from `.ci_support/*`. What's going on?

@conda-forge-admin, please rerender

**Comment by dalcinl** [comment]
Oh, sorry, you removed builds to speedup things.

Regarding the Fortran compiler version,  you may be wrong. Look in the various files in `.ci_support/`, they are setup to use version 12. Don't get confused with the libgfortran packages.

**Comment by dalcinl** [comment]
Damn.. if you look carefully at the logs, then you see mixed versions of the packages. For example:

```
2024-01-31T20:14:24.6549910Z     gfortran_impl_osx-64:        13.2.0-h2bc304d_2         conda-forge
2024-01-31T20:14:24.6550370Z     gfortran_impl_osx-arm64:     12.3.0-h4f4d140_2         conda-forge
2024-01-31T20:14:24.6550820Z     gfortran_osx-arm64:          12.3.0-h64a2375_1         conda-forge
```
Maybe our issue is related to cross-compiling, the header is not there where `$FC -print-search-dirs` says, but elsewhere.

File: recipe/build-mpi.sh [hunk 3]
```diff
@@ -33,7 +33,29 @@ if [[ "$target_platform" == osx-* ]]; then
   # Add gfortran internal header to clang include dir
   fcdir=$($FC -print-search-dirs | awk '/install: /{print $2}')
   ccdir=$($CC -print-search-dirs | awk '/libraries: =/{print substr($2,2)}')
-  ln -s $fcdir/include/ISO_Fortran_binding.h $ccdir/include
+  echo "=== DEBUG ==="
```
Comment (dalcinl) [inline recipe/build-mpi.sh:36]:
```suggestion
  echo "=== DEBUG ==="
  echo "ls $fcdir"
  ls $fcdir
  echo "ls $fcdir/include"
  ls $fcdir/include
```

**Comment by aradi** [comment]
@dalcinl I am unfortunately (as you might have note already :wink:) still a newbie in Conda. But, the compiler fix https://github.com/conda-forge/ctng-compilers-feedstock/pull/123 which was needed in order to make the PR for the MPI-F08 interface (#84) work at all: does this affect any 12.3 compiler packages? And does it affect any osx compiler packages at all (I see only linux-builds in their CI)? Can it be, that on the osx_arm64 cross-compiling case, the `ISO_Fortran_binding.h` is not installed at all, yet?

**Comment by dalcinl** [comment]
-

> @dalcinl I am unfortunately (as you might have note already 😉) still a newbie in Conda.

Me too! And my frustration grows every day...

> But, the compiler fix [conda-forge/ctng-compilers-feedstock#123](https://github.com/conda-forge/ctng-compilers-feedstock/pull/123) which was needed in order to make the PR for the MPI-F08 interface (#84) work at all: does this affect any 12.3 compiler packages? 

I'm not really sure. If you manually install these packages in bare metal, then the header is definitely there.

> And does it affect any osx compiler packages at all (I see only linux-builds in their CI)? Can it be, that on the osx_arm64 cross-compiling case, the `ISO_Fortran_binding.h` is not installed at all, yet?

In the cross-compiling case, the header is definitely NOT there, look at the `ls` output. This probably means that somehow a previous build of the gfortran package is being picked up.

Also, look at the logs:


```
## Package Plan ##

  environment location: /Users/runner/miniforge3/conda-bld/mpich-mpi_1706764906406/_h_env_placehold_


The following NEW packages will be INSTALLED:

    ...
    gfortran_impl_osx-arm64:     12.3.0-h53ed385_2         conda-forge
    gfortran_osx-arm64:          12.3.0-h57527a5_1         conda-forge

...

## Package Plan ##

  environment location: /Users/runner/miniforge3/conda-bld/mpich-mpi_1706764906406/_build_env


The following NEW packages will be INSTALLED:

    ...
    gfortran_impl_osx-arm64:     12.3.0-h4f4d140_2         conda-forge
    gfortran_osx-arm64:          12.3.0-h64a2375_1         conda-forge
    ...
```
Do you notice that in these two cases the build string hashes are not the same?

Also note what we have in our `meta.yaml`:

```yaml
      host:
        - {{ fortran_compiler }}_{{ target_platform }} {{ fortran_compiler_version }}  # hack to force conda to make both fortran abi builds
```
and I have no idea what that is for. Hacks and a temporary workarounds that gets added on top of previous hacks and workarounds, things never get consolidated and cleaned up. Therefore my frustration... 

Maybe we should try to remove that stuff from our `meta.yaml` and see what happen?

**Comment by aradi** [comment]
@conda-forge-admin, please rerender

**Comment by aradi** [comment]
> Maybe we should try to remove that stuff from our `meta.yaml` and see what happen?

Tried, does not help. :disappointed:

**Comment by aradi** [comment]
OK, so not sure where to go from here...

As a side note: In my desperation, I even tried to install on my Linux (!) system `gfortran_impl_osx-64` and `gfortran_impl_osx-arm64` (version 13.2), but none of them contained the include file. However, I am not sure, whether this has any significance...

**Comment by dalcinl** [comment]
OK, something is definitely not working as we expect. And I believe the issue is related to different package versions/build  depending on the architecture:

1) If you do 
```
$ env CONDA_SUBDIR=osx-64 micromamba install gfortran_impl_osx-arm64
```
then you get the following package build:
```
  + gfortran_impl_osx-arm64   13.2.0  hbfbf36e_1  conda-forge      19MB
```
and this package DOES NOT install `ISO_Fortran_binding.h`:
```
ls $CONDA_PREFIX/lib/gcc/arm64-apple-darwin20.0.0/13.2.0/include/ISO_Fortran_binding.h
ls: cannot access '<envdir>/lib/gcc/arm64-apple-darwin20.0.0/13.2.0/include/ISO_Fortran_binding.h': No such file or directory
```

2) If you do 
```
$ env CONDA_SUBDIR=osx-arm64 micromamba install gfortran_impl_osx-arm64
```
then you get the following package build:
```
  + gfortran_impl_osx-arm64       13.2.0  h30f4408_1  conda-forge      18MB
```
and this package DOES install `ISO_Fortran_binding.h`:
```
ls $CONDA_PREFIX/lib/gcc/arm64-apple-darwin20.0.0/13.2.0/include/ISO_Fortran_binding.h
<envdir>/lib/gcc/arm64-apple-darwin20.0.0/13.2.0/include/ISO_Fortran_binding.h
```

Therefore, you either get or not get the header depending on whether you are or are not cross-compiling. This is a mess I'm not prepared to deal with. At this point I don't know what's the proper way to address this issue and move forward. I guess the only option is to ask the conda-forge folks via chat https://app.gitter.im/#/room/#conda-forge_conda-forge-compilers:gitter.im.

**Comment by h-vetinari** [comment]
Gfortran for OSX is built on a different than https://github.com/conda-forge/ctng-compilers-feedstock. I know it's not very transparent where packages come from sometimes, and this is double true for the compiler packages. But those are also just very complicated to bundle, so I recommend the search under https://conda-forge.org/packages/, which will show you which particular feedstock produces a given output.

In this case, it's very likely that we need an equivalent of https://github.com/conda-forge/ctng-compilers-feedstock/pull/123 also for https://github.com/conda-forge/gfortran_osx-64-feedstock

**Comment by aradi** [comment]
@h-vetinari Thx, I've opened an appropriate issue there!

**Comment by aradi** [comment]
@conda-forge-admin, restart ci

**Comment by aradi** [comment]
@conda-forge-admin please rerender

**Comment by aradi** [comment]
As far as I can tell (the builds are still running), the logs on all 5 platforms contain `checking for Fortran 2008 support... yes`, so the `mpi_f08` interface should now built everywhere

**Comment by aradi** [comment]
@dalcinl :tada: I can now build my MPI-project on all platforms. Thanks for your help to bring that forward.

**Pull request merged**